cloud-api/src/main/java/com/sequenceiq/cloudbreak/cloud/NetworkConnector.java

```diff
@@ -1,26 +1,84 @@
 package com.sequenceiq.cloudbreak.cloud;
 
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
+import java.util.Random;
 
 import com.sequenceiq.cloudbreak.cloud.model.CloudCredential;
 import com.sequenceiq.cloudbreak.cloud.model.CloudSubnet;
 import com.sequenceiq.cloudbreak.cloud.model.Network;
 import com.sequenceiq.cloudbreak.cloud.model.SubnetSelectionParameters;
+import com.sequenceiq.cloudbreak.cloud.model.SubnetSelectionResult;
 import com.sequenceiq.cloudbreak.cloud.model.network.CreatedCloudNetwork;
 import com.sequenceiq.cloudbreak.cloud.model.network.NetworkCreationRequest;
 import com.sequenceiq.cloudbreak.cloud.model.network.NetworkDeletionRequest;
-import com.sequenceiq.cloudbreak.cloud.model.SubnetSelectionResult;
 
 /**
  * Network connectors.
  */
 public interface NetworkConnector extends CloudPlatformAware {
 
+    String NOT_ENOUGH_AZ = "Acceptable subnets are in %d different AZs, but subnets in %d different AZs required.";
+
     CreatedCloudNetwork createNetworkWithSubnets(NetworkCreationRequest networkCreationRequest);
 
     void deleteNetworkWithSubnets(NetworkDeletionRequest networkDeletionRequest);
 
     String getNetworkCidr(Network network, CloudCredential credential);
 
-    SubnetSelectionResult selectSubnets(List<CloudSubnet> subnetMetas, SubnetSelectionParameters subnetSelectionParameters);
+    SubnetSelectionResult filterSubnets(Collection<CloudSubnet> subnetMetas, SubnetSelectionParameters subnetSelectionParameters);
+
+    int subnetCountInDifferentAzMin();
+
+    int subnetCountInDifferentAzMax();
+
+    default SubnetSelectionResult selectSubnets(SubnetSelectionResult subnetSelectionResult, SubnetSelectionParameters subnetSelectionParameters) {
+        if (subnetSelectionResult.hasResult()) {
```

Comment:
We should add a lot more logging to figure out what happens when it's not something that we expect.

Reply:
yep already mentioned in the PR description